### PR TITLE
Debug heuristics

### PR DESCRIPTION
**Pull request by szschaler**
Support for debug heuristics allowing to restrict the set of steps shown in the logical step decider.

Todo:

- [x] Setup basic generic code structure

- [ ] Implement actual heuristics
  - Concurrency heuristics
     - [x] Only parallelise matches of specific sets of rules -- focus in on issues with concurrency of a particular rule / rule set
    - [x] Only parallelise steps that overlap in non-changing parts (focus on concurrency over a shared resource -- highlights where the same machine is used concurrently)
    - [ ] Only parallelise steps that don't overlap in non-changing parts (focus on concurrency over a shared resource -- assume machines cannot be used concurrently for more than one job, without encoding mutual exclusion in the operational semantics – this one might be useful for model checking as well as a kind of computational model)
     - [ ] Only parallelise steps that overlap in specific objects (or objects of a specific type) -- remove concurrent generation of new parts, but keep all other processing activities

  - Filtering heuristics
     - [x] Only parallelise a specific maximum number of rule matches -- generic notion of limited concurrency
     - [x] Remove steps that differ only in objects of a specific type -- specify that objects of that type are not considered to have identity (e.g., parts). Note, however, how this needs to interact with concurrency computation to recognise that it is still meaningful to have two potentially concurrent applications of assemble, for example. It's just not meaningful to have four ``different'' atomic assemble steps where there is only one machine. Hence, this is a filtering heuristic that needs to be applied after all possible concurrent executions have been computed

- [ ] Implement UI to select heuristics
  - [x] From launch tabs
  - [ ] During debugging

### CODE REVIEW

**Comment by kbojarczuk** [comment]
I've implemented the stuff I have ticked off on the list above and I have tested it and it seems to work but there is an issue with that 'fold' function you have implemented. When I use the max steps heuristics and call it the normal way(I left it as commented out and the code I used for testing too) then it works but with the fold it doesn't(it just doesn't apply the changes). It'd be great if you could have a look at that 'fold' as you're more familiar with the syntax. If it doesn't work then I can just implement it without the fold on Tue.

**Comment by szschaler** [comment]
Fixed :-)

File: uk.ac.kcl.inf.modelling.xdsml.gemoc_henshin/src/uk/ac/kcl/inf/modelling/xdsml/gemoc_henshin/engine/core/HenshinStep.xtend [hunk 1]
```diff
@@ -15,8 +15,8 @@ import org.eclipse.gemoc.trace.commons.model.trace.TracePackage
  */
 class HenshinStep extends GenericSmallStepImpl {
 
-	protected Match match
-	protected List<Match> matches
+	public Match match
```
Comment (szschaler) [inline uk.ac.kcl.inf.modelling.xdsml.gemoc_henshin/src/uk/ac/kcl/inf/modelling/xdsml/gemoc_henshin/engine/core/HenshinStep.xtend]:
```suggestion
        @Accessors(PUBLIC_GETTER)
	private Match match
```

File: uk.ac.kcl.inf.modelling.xdsml.gemoc_henshin/src/uk/ac/kcl/inf/modelling/xdsml/gemoc_henshin/engine/core/HenshinStep.xtend [hunk 2]
```diff
@@ -15,8 +15,8 @@ import org.eclipse.gemoc.trace.commons.model.trace.TracePackage
...
+	public List<Match> matches
```
Comment (szschaler) [inline uk.ac.kcl.inf.modelling.xdsml.gemoc_henshin/src/uk/ac/kcl/inf/modelling/xdsml/gemoc_henshin/engine/core/HenshinStep.xtend]:
```suggestion
        @Accessors(PUBLIC_GETTER)
	private List<Match> matches
```

**Comment by kbojarczuk** [comment]
I've pushed my code and i've tested it, it all seems to work. I've tried to implement the registry you've mentioned but it makes it much more confusing? I'm not sure there's a point in it as it just seems to overcomplicate things as you've mentioned passing IDs to the solver rather than values? Later we will have to add an integer to the maxSteps heuristics anyway and that will complicate it even more? Most of the stuff still has to be hardcoded anyway and there's not that many heuristics so I've implemented it like this for now.

**Comment by szschaler** [comment]
I've added the registry that I had in mind. This will allow us to later switch to an extension-based setup. It also means there is only one place where the list of heuristics and the IDs are kept and we don't have to duplicate them.

Next steps (could you look into these?): 

- [ ] Add a flag to a `HeuristicDefinition` telling whether it describes a filter or concurrency heuristic. Use this to allocate checkboxes to groups in the launch tab and to add generated heuristics to the solver (rather than the current `instanceof` check
- [ ] Add my new heuristic to the set of heuristics
- [ ] Add support for configurable heuristics: `HeuristicDefinition` needs:
  - [ ] an operation to generate additional controls for the launch tab
  - [ ] an operation to take the contents of these controls and store them in a configuration
  - [ ] an extension to the current `instantiate` method that takes a configuration as a parameter and extracts any additional information needed to pass as parameters to the heuristic once instantiated

**Comment by szschaler** [comment]
Split out remaining work into separate issues: #26 and #27

**Pull request merged**